feat(API): Add descendant find for parent db objects.

d-bohls opened the pull request:
Fixes #239

The following parent database objects now have a find method:
- database
- cluster
- lin schedule
- pdu
- frame
- subframe

All database object are now subclasses of
DatabaseObject, which is used as an interface.
Database objects are now exposed through the
nixnet.database namespace, and their constructors
now take kwargs to signal to users that these
objects should not be instantiated by users.

I tested on the shipping example database "NIXNET_example"
to ensure I can indeed find children and grandchildren,
and that sane errors are thrown when the object name
not found or the object class is wrong.
Unit tests will be added in #238.

- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nixnet-python/blob/master/CONTRIBUTING.rst).
- [ ] ~~New tests have been created for any new features or regression tests for bugfixes.~~
- [x] `tox` successfully runs, including unit tests and style checks (see [CONTRIBUTING.md](https://github.com/ni/nixnet-python/blob/master/CONTRIBUTING.rst)).

d-bohls commented:
A couple of notes on this review.

After #237 goes in, I plan to add docstrings to this review, because the docstrings for this review will reference docstrings in the other review.

I am deferring unit tests for this review to be done as part of #238.

epage commented:
Would it be possible to post a comment of what your modified example looks like, so we can see what the API is like to use?  Or more ideally, tests :)

epage commented on nixnet/database/_find_object.py:
```diff
@@ -0,0 +1,52 @@
+        _errors.check_for_error(_cconsts.NX_ERR_INVALID_PROPERTY_VALUE)
+        # The above line will raise an exception.
+        # However, mypy doesn't know that and will complain if the following return is missing.
+        return None
```
Looks like you could make the return type a union of the regular and [`NoReturn`](https://mypy.readthedocs.io/en/latest/kinds_of_types.html#the-noreturn-type)

d-bohls commented on nixnet/database/_find_object.py:
```diff
@@ -0,0 +1,52 @@
+
+
+def find_object(parent_handle, object_class, object_name):
+    # type: (int, constants.ObjectClass, typing.Text) -> object
```
Regarding 1, I had thought about doing it that way. I guess on the plus side, it eliminates an enum the user has to deal with, and encourages familiarity with the actual objects. On the con side, those of us who use IDEs lose the benefit of intellisense for the enum values. I think my preference would be to continue using the enum as I don't really see much benefit to passing in an object instead of an enum. I'm open to either way though.

epage replied:
> On the con side, those of us who use IDEs lose the benefit of intellisense for the enum values. 

Completion for `nixnet.constants.ObjectClass` is not much different than completion for `nixnet.database.`

As for other intellisense features, you can improve that by
1. Providing a common base class for all database objects
2. Change the annotations from `type: (int, constants.ObjectClass, typing.Text) -> object` to `type: (int, typing.Type[nixnet.database.SomeBaseClass], typing.Text) -> nixnet.database.SomeBaseClass`

This is similar to the approach we took elsewhere in the API
- Define [`FrameFactory`](https://github.com/ni/nixnet-python/blob/master/nixnet/types.py#L248) have have each frame implement it
- Accept a frame object when [reading](https://github.com/ni/nixnet-python/blob/7f79f5b3a23c5ffda8f965fcd2d03ad2316f8b88/nixnet/_session/frames.py#L97) and [converting](https://github.com/ni/nixnet-python/blob/e9a946b035db388030b13c426a58f01a5afe4de2/nixnet/convert.py#L216)

> I think my preference would be to continue using the enum as I don't really see much benefit to passing in an object instead of an enum.

Advantages
- Minimize vocabulary terms the user of the API has to deal with.  I need a `Cluster`, so let me pass a `Cluster` vs having to deal with an enum that represents `Cluster`.
- Avoid C-isms.  Part of the role of the enum is a form of RTTI in C.  Changing it to the actual python class is exposing the same concept in the python way
- Controlled access.  The doc string and intellisense will help guide people to the correct values to pass in rather than having things like `System` be an option.

epage commented:
For reference, with using classes instead of enums

```python
import nixnet
from nixnet import database


def main():
    with database.Database("NIXNET_example") as db:
        cluster = db.find(database.Cluster, "CAN_Cluster")
        print(cluster.name)
        frame1 = db.find(database.Frame, "CANCyclicFrame2")
        frame2 = cluster.find(database.Frame, "CANCyclicFrame2")
        print(frame1.name)
        print(frame1.name)
        assert (frame1 == frame2)
        signal = db.find(database.Signal, "ClutchPressure")
        print(signal.name)
```

Oh, fun, that won't work as of this moment.  `nixnet.database` only exposes `Database` and not the other types
https://github.com/ni/nixnet-python/blob/master/nixnet/database/__init__.py

Interesting.  I know the concern was about exposing the useless private constructors to users but this doesn't allow our users to use mypy.

A workaround for exposing the useless private constructors is to make `handle` a keyword-only arg that is prefixed with `_`.  That isn't officially supported in python2 but I recently learned about a backport of the feature: https://pypi.python.org/pypi/kwonly-args

d-bohls commented:
I'm curious about your reasoning for suggesting kwonly-args and not kwargs? Won't both work fine for this purpose? kwonly-args looks like it might be a cleaner implementation.

epage commented:
> I'm curious about your reasoning for suggesting kwonly-args and not kwargs? Won't both work fine for this purpose? kwonly-args looks like it might be a cleaner implementation.

If you are taking about usage, any keyword argument can also be a positional argument.

If you are talking about `**kwargs`, then that is effectively another way of defining a keyword-only argument and fits the spirit of what I'm suggesting.  The literal keyword-only argument feature just requires less boiler plate / is more clear on intent.

d-bohls commented:
I've done the following:

1. Introduced DatabaseObjectBase for all database object to inherit from
2. Exposed them in database package's init
3. Modified find methods to use DatabaseObjectBase objects

However, I'm still not seeing a useful namespacing experience. Not sure yet what I'm missing.

I'm holding off on producing docstrings and making the kwargs constructor changes until I can fix the namespacing experience.

epage commented on nixnet/database/_find_object.py:
```diff
@@ -26,28 +26,31 @@ def find_object(
+        class_enum = constants.ObjectClass.SUBFRAME
+    else:
+        # arbitrary value for class_enum so mypy sees a value in each condition
+        class_enum = constants.ObjectClass.CLUSTER
```
I'd actually recommend a `raise ValueError("Unsupported database object")`

epage replied:
Oh, missed the raise below :)

epage commented on nixnet/database/__init__.py:
```diff
@@ -3,7 +3,24 @@
 
 
-__all__ = ["Database"]
+__all__ = [
```
`DatabaseObjectBase` should be in here so users can do mypy typing

d-bohls replied:
done

epage commented on nixnet/database/_database_object_base.py:
```diff
@@ -0,0 +1,31 @@
+import typing  # NOQA: F401
+
+
+class DatabaseObjectBase(object):
```
Since this should be public, we should ensure we have a name we like.  I personally find -`Base` suffixes a little off, at least in python code.

d-bohls replied:
I was following what I saw with [SessionBase](https://github.com/ni/nixnet-python/blob/master/nixnet/_session/base.py#L22). So your suggestion is `DatabaseObject`?

epage replied:
iirc SessionBase was me doing what I said not to do, concrete inheritance for code sharing.  Its not actually exposed in the API.

Something like `DatabaseObject` might work.

epage commented:
> I've done the following:

Looks slick

> However, I'm still not seeing a useful namespacing experience. Not sure yet what I'm missing.

What do you mean?

d-bohls commented:
@epage 

> However, I'm still not seeing a useful namespacing experience. Not sure yet what I'm missing.
> What do you mean?

Apologies for not be clear there. I was talking about the intellisense issue where I don't see anything useful in the database namespace after typing "database.".

Turns out, this seems to be a PyCharm quirk where my nixnet-python package was defined in my PyCharm project (rather than being installed traditionally) and using a loose python script outside of the project to test with. It's strange because the script would run just fine, but I wasn't getting any intellisense for things in nixnet. Once I include my test files as a source root in the project, the intellisense automagically works.

d-bohls commented:
Added docstrings and modified database objects to init with kwargs.

@jashnani and I chose to not include docs for the new DatabaseObject base class.

Thanks to @jashnani and @epage for not settling on "good enough". I really like the final design.

I'll add tests for this and all database objects next.

d-bohls commented:
@jashnani Just in case, let me squash before merge.

d-bohls commented:
The only "less than ideal" issue I might mention is that the `DatabaseObject` base class shows up in intellisense, though this is better than the enum alone since other things like `Session` showed up in that.

![image](https://user-images.githubusercontent.com/33291082/39288413-777f6736-48ed-11e8-8acc-8a4904a8796a.png)

jashnani commented on nixnet/database/_database_object.py:
```diff
@@ -0,0 +1,41 @@
+        return hash(self._handle)
+
+    def __repr__(self):
+        return '{}(handle={})'.format(type(self).__name__, self._handle)
```
Our use of DatabaseObject is meant for interface only. I would've expected these implementations to continue living in child classes and evolve as necessary:

`__eq__`
`__ne__`
`__hash__`
`__repr__`

d-bohls replied:
I read the information you sent me on this. I do understand some of the benefits mentioned to not DNR (like code locality). In this particular situation, I think I'd still prefer not to duplicate theses four functions in all eight subclasses. Regardless, I've moved them back as they were and I'll add unit tests as appropriate in my upcoming PR.

jashnani commented on nixnet/database/_signal.py:
```diff
@@ -483,19 +468,19 @@ def mux_value(self):
     def mux_subfrm(self):
         # actually returns _subframe.SubFrame, but avoiding a circular import
-        # type: () -> typing.Any
+        # type: () -> _database_object.DatabaseObject
```
Should the return type just be SubFrame? Ditto as above.

d-bohls replied:
I was able to get this to work using the `MYPY = False` technique.

d-bohls commented on nixnet/database/_signal.py:
```diff
@@ -187,14 +172,14 @@ def default(self, value):
     def frame(self):
         # actually returns _frame.Frame, but avoiding a circular import
-        # type: () -> typing.Any
+        # type: () -> _database_object.DatabaseObject
```
> Shouldn't the return type just be Frame?

Yes, that's what the comment is saying. I couldn't find a way to import before the type check, and not for lack of trying.

jashnani replied:
There's a way to do it by setting MYPY = false. See details here: https://mypy.readthedocs.io/en/latest/common_issues.html#import-cycles

I'm not going to hold the review for this, so don't feel like you have to go out of way to do this.

d-bohls replied:
I was able to get this to work using `MYPY = false`. Cool.

d-bohls commented on nixnet/database/_database_object.py:
```diff
@@ -0,0 +1,41 @@
+    ):
+        # type: (...) -> None
+        if not kwargs or '_handle' not in kwargs:
+            _errors.raise_xnet_error(_cconsts.NX_ERR_INVALID_PROPERTY_VALUE)
```
You mentioned this PEP for keyword-only arguments:

https://www.python.org/dev/peps/pep-3102/ 

"In accordance with the current Python implementation, any errors encountered will be signaled by raising TypeError."

Changed to raise TypeError(),

d-bohls commented on nixnet/database/_find_object.py:
```diff
@@ -0,0 +1,54 @@
+        class_enum = constants.ObjectClass.SIGNAL
+    elif object_class is SubFrame:
+        class_enum = constants.ObjectClass.SUBFRAME
+    else:
```
I had thought about using a dictionary but had no strong preference.

It's a dictionary now. If the key isn't found, I'm raising a ValueError for the object_class argument.

d-bohls commented on nixnet/database/_collection.py:
```diff
@@ -58,7 +58,7 @@ def __getitem__(self, index):
         if isinstance(index, six.string_types):
             ref = _funcs.nxdb_find_object(self._handle, self._type, index)
-            return self._factory(ref)
+            return self._factory(_handle=ref)
```
Done

d-bohls commented on nixnet/database/_subframe.py:
```diff
@@ -141,7 +177,7 @@ def name(self, value):
     def pdu(self):
         # actually returns _pdu.Pdu, but avoiding a circular import
-        # type: () -> typing.Any
+        # type: () -> _database_object.DatabaseObject
```
I was able to get this to work using the `MYPY = False` technique.
